optuna/samplers/grid.py

```diff
@@ -108,22 +111,29 @@ def sample_relative(self, study, trial, search_space):
         # object is hard to get at the beginning of trial, while we need the access to the object
         # to validate the sampled value.
 
-        unvisited_grids = self._get_unvisited_grid_ids(study)
+        target_grids = self._get_unvisited_grid_ids(study)
 
-        if len(unvisited_grids) == 0:
-            raise ValueError(
-                "All grids have been evaluated. If you want to avoid this error, "
-                "please make sure that unnecessary trials do not run during "
-                "optimization by properly setting `n_trials` in `study.optimize`."
+        if len(target_grids) == 0:
+            _logger.warning(
+                "`GridSampler` is evaluating a duplicated point because all grids "
+                "have been evaluated. This may happen due to a timing issue during "
+                "distributed optimization or an unnecessary number of `n_trials`."
             )
 
+            study.stop()
+            target_grids = list(range(len(self._all_grids)))
+
         # In distributed optimization, multiple workers may simultaneously pick up the same grid.
         # To make the conflict less frequent, the grid is chosen randomly.
-        grid_id = random.choice(unvisited_grids)
+        grid_id = random.choice(target_grids)
 
         study._storage.set_trial_system_attr(trial._trial_id, "search_space", self._search_space)
         study._storage.set_trial_system_attr(trial._trial_id, "grid_id", grid_id)
 
+        # Once all grids are picked up, optimization stops after the current trial finishes.
+        if len(self._get_unvisited_grid_ids(study, include_unfinished=True)) == 0:
```

Comment:
I'm a bit worried that it's not immediately clear why `_get_unvisited_grid_ids` is called twice in this method and with different arguments. At least it isn't to me. Would you mind leaving a comment about this logic?

Reply:
Thanks for your pointing this out. As you mention, the duplicated call of this method seems unnecessary. I just simplified logic.